File: src/core/responder/focus.ts
import { Controller } from "../view";
import {
  type Responder,
  _setFirstResponder,
  currentFirstResponder,
} from "./responder";
import { View } from "../view";

/**
 * The responder that should get first chance for a DOM event/focus
 * location. This is the DOM-to-responder adapter: raw elements walk up to
 * their first managed View; controller root Views yield their controller,
 * while control/outlet Views yield the View itself.
 */
export function firstResponderFor(el: HTMLElement | null): Responder | null {
  let node = el;
  while (node) {
    const v = View.of(node);
    if (v) return v.controller ?? v;
    node = node.parentElement;
  }
  return null;
}

export function responderElement(r: Responder): HTMLElement | null {
  if (r instanceof Controller) return r.view.element;
  if (r instanceof View) return r.element;
  return null;
}

export function syncFirstResponderFromFocus(
  target: HTMLElement | null,
  previouslyFocused: HTMLElement | null = null,
): void {
  const responder = firstResponderFor(target);
  const current = currentFirstResponder();
  if (current === responder) return;

  // Focus-DRIVEN path: the browser already moved DOM focus. We only
  // reconcile the FR singleton, and still honor a resign-refusal.
  if (current && !current.resignFirstResponder(responder)) {
    const ownerEl = responderElement(current);
    if (ownerEl) {
      let restore: HTMLElement | null = null;
      if (previouslyFocused && ownerEl.contains(previouslyFocused)) {
        restore = previouslyFocused;
      } else {
        restore = View.of(ownerEl)?.firstFocusableElement() ?? null;
      }
      if (restore) restore.focus();
    }
    return;
  }
  _setFirstResponder(responder);
}

File: opal/lib/swill/focus.rb
# backtick_javascript: true

module Swill
  # The DOM-to-responder adapter: which responder owns a DOM location, and how
  # to reconcile the first-responder singleton when the browser moves focus on
  # its own (a click or Tab into a field).
  module Focus
    module_function

    # The responder that should get first chance at an event/focus at +element+.
    # Walk up to the first managed View: a controller's root View yields its
    # controller; a control/outlet View yields itself.
    def responder_for(element)
      node = element
      while node
        view = View.for(node)
        return (view.controller || view) if view

        node = `#{node}.parentElement`
      end
      nil
    end

    def responder_element(responder)
      return responder.view.element if responder.is_a?(Controller)
      return responder.element if responder.is_a?(View)

      nil
    end

    # Reconcile the FR after the browser has already moved DOM focus. Honors a
    # resign-refusal by restoring focus to the outgoing responder; otherwise
    # records the new FR without re-running `become` (the DOM is already there).
    def sync_from_focus(target, previously_focused = nil)
      responder = responder_for(target)
      current = FirstResponder.current
      return if current.equal?(responder)

      if current && !current.resign_first_responder(responder)
        restore_focus(current, previously_focused)
        return
      end

      FirstResponder.install(responder)
    end

    def restore_focus(responder, previously_focused)
      owner = responder_element(responder)
      return unless owner

      restore =
        if previously_focused && `#{owner}.contains(#{previously_focused})`
          previously_focused
        else
          view = View.for(owner)
          view&.first_focusable_element
        end

      `#{restore} && #{restore}.focus()` if restore
    end
  end
end
